apps/gateway/src/middlewares/rateLimit.ts

```diff
@@ -0,0 +1,52 @@
+import type { ApiKey } from '@latitude-data/core/browser'
+import { buildRedisConnection } from '@latitude-data/core/redis'
+import { env } from '@latitude-data/env'
+import { RateLimiterRedis, RateLimiterRes } from 'rate-limiter-flexible'
+import { createMiddleware } from 'hono/factory'
+import { RateLimitError } from '@latitude-data/core/lib/errors'
+
+
+declare module 'hono' {
+    interface ContextVariableMap {
+        apiKey: ApiKey
+    }
+}
+
+
+const RATE_LIMIT_POINTS = 1000
+const RATE_LIMIT_DURATION = 60
+
+const rateLimiter = new RateLimiterRedis({
+    storeClient: await buildRedisConnection({
+        host: env.QUEUE_HOST,
+        port: env.QUEUE_PORT,
+        password: env.QUEUE_PASSWORD,
+        enableOfflineQueue: true,
+        maxRetriesPerRequest: null,
+    }),
+    points: RATE_LIMIT_POINTS,
+    duration: RATE_LIMIT_DURATION,
+});
+
+const rateLimitMiddleware = () =>
+    createMiddleware(async (c, next) => {
+        const workspaceId: number = c.get('workspace').id
```

Comment:
hmm instead of doing it by workspace id we could do it by api key and put this middleware before the auth middleware so that we avoid unnecessary db rountrips to get the workspace before immediately rate limiting the user